File: vidyapath/server/models/School.js
const mongoose = require('mongoose');

const schoolSchema = new mongoose.Schema({
  udiseCode: { type: String, unique: true, sparse: true, trim: true },
  name: { type: String, required: true, trim: true },
  board: { type: String, enum: ['CBSE', 'ICSE', 'State', 'IB', 'IGCSE', 'Other'] },
  type: { type: String, enum: ['government', 'private', 'aided'] },
  address: {
    city: String, district: String, taluka: String, state: String, pincode: String,
    fullAddress: String,
  },
  location: {
    type: { type: String, enum: ['Point'], default: 'Point' },
    coordinates: { type: [Number], index: '2dsphere' }, // [longitude, latitude]
  },
  contact: {
    email: { type: String, default: 'contact@school.edu' },
    phone: { type: String, default: '' },
    website: { type: String, default: '' }
  },
  facilities: {
    hasComputerLab: { type: Boolean, default: false },
    hasLibrary: { type: Boolean, default: false },
    hasPlayground: { type: Boolean, default: false },
    hasSmartClasses: { type: Boolean, default: false },
  },
  stats: {
    totalStudents: Number,
    studentTeacherRatio: Number,
    avgPassPercentage: Number,
  },
  ratings: {
    overall: { type: Number, default: 0 },
    academics: { type: Number, default: 0 },
    infrastructure: { type: Number, default: 0 },
    faculty: { type: Number, default: 0 },
    extracurricular: { type: Number, default: 0 },
    safety: { type: Number, default: 0 },
    communication: { type: Number, default: 0 },
    valueForMoney: { type: Number, default: 0 },
    totalReviews: { type: Number, default: 0 },
  },
  customFields: { type: Map, of: mongoose.Schema.Types.Mixed, default: {} },
  isVerified: { type: Boolean, default: false },
  adminUsers: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User' }],
  updateToken: { type: String, sparse: true, index: true },
  updateTokenExpires: { type: Date },
  updateRequestStatus: { type: String, enum: ['none', 'sent', 'visited', 'updated'], default: 'none' },
}, { timestamps: true });

schoolSchema.index({ 'address.state': 1, 'address.district': 1 });

const bookmarkSchema = new mongoose.Schema({
  userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  opportunityId: { type: mongoose.Schema.Types.ObjectId, ref: 'Opportunity', required: true },
}, { timestamps: true });

bookmarkSchema.index({ userId: 1, opportunityId: 1 }, { unique: true });

const School = mongoose.model('School', schoolSchema);
const Bookmark = mongoose.model('Bookmark', bookmarkSchema);

module.exports = { School, Bookmark };

File: vidyapath_app/lib/models/school.dart
class SchoolModel {
  final String id;
  final String name;
  final String? board;
  final String? type;
  final Map<String, dynamic>? address;
  final Map<String, dynamic>? contact;
  final Map<String, dynamic>? facilities;
  final Map<String, dynamic>? stats;
  final Map<String, dynamic>? ratings;
  final bool isVerified;
  final List<String>? courses;
  final String? description;
  final String? affiliation;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  SchoolModel({
    required this.id,
    required this.name,
    this.board,
    this.type,
    this.address,
    this.contact,
    this.facilities,
    this.stats,
    this.ratings,
    this.isVerified = false,
    this.courses,
    this.description,
    this.affiliation,
    this.createdAt,
    this.updatedAt,
  });

  factory SchoolModel.fromJson(Map<String, dynamic> json) {
    return SchoolModel(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      board: json['board'],
      type: json['type'],
      address: json['address'] is Map ? Map<String, dynamic>.from(json['address']) : null,
      contact: json['contact'] is Map ? Map<String, dynamic>.from(json['contact']) : null,
      facilities: json['facilities'] is Map ? Map<String, dynamic>.from(json['facilities']) : null,
      stats: json['stats'] is Map ? Map<String, dynamic>.from(json['stats']) : null,
      ratings: json['ratings'] is Map ? Map<String, dynamic>.from(json['ratings']) : null,
      isVerified: json['isVerified'] ?? false,
      courses: json['courses'] is List ? List<String>.from(json['courses']) : null,
      description: json['description'],
      affiliation: json['affiliation'],
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.tryParse(json['updatedAt']) : null,
    );
  }

  String get city => address?['city'] ?? '';
  String get state => address?['state'] ?? '';
  String get district => address?['district'] ?? '';
  String get email => contact?['email'] ?? '';
  String get phone => contact?['phone'] ?? '';
  String get website => contact?['website'] ?? '';
  double get overallRating => (ratings?['overall'] as num?)?.toDouble() ?? 0.0;
  int get totalReviews => (ratings?['totalReviews'] as num?)?.toInt() ?? 0;
  bool get hasComputerLab => facilities?['hasComputerLab'] == true;
  bool get hasLibrary => facilities?['hasLibrary'] == true;
  bool get hasPlayground => facilities?['hasPlayground'] == true;
  bool get hasSmartClasses => facilities?['hasSmartClasses'] == true;
}
